### PR TITLE
devices: Generating device list

### PR DESCRIPTION
**Pull request by garethtdavies**
Update: This PR now just includes everything to correct the current supported device listings. Let's get this page in a decent state before we aim to display this dynamically.

---

Here is a first pass at generating this page dynamically: https://www.balena-staging.io/docs/reference/hardware/devices/ which is very out of date. This is generated directly from https://api.balena-cloud.com/device-types/v1/

Staged here: https://www.balena-staging.io/docs/reference/hardware/devices/
Script used to generate is here which can be used to generate the partial: https://gist.github.com/garethtdavies/486e37df0b8491afeabb8dd82b6098b4

I've obviously changed the layout of the page away from using the device images (we don't have them all) and there isn't any grouping. I've used icons from https://github.com/balena-io/balenacloud-dashboard/tree/6848d0759e47929c6f99f5dca081327b34c2b977/src/img/devices and replaced all existing for consistency which are pulled based on the `slug` property. 

Just updating here means that the Getting Started guide won't have them all. Clearly having all of these in the Getting Started guide should be the goal but that will take a bit more work due to the way that is structured. So this might be the best first step in updating the device docs in a manageable way.

~I'm not sure if we want all that info in the table (Machine Name, Arch, balenaOS Version). While balenaOS version is a nice addition this does raise the issue of keeping this up to date as, as written this is manually run. Even if it was scripted to run on deploy, we'd need a way to trigger (ref other discussions e.g. Masterclass PR). Maybe we simply remove for now so we can deploy this and find a way to automatically refresh this.~

Change-type: minor
Connects-to: #1058 
Signed-off-by: Gareth Davies <gareth@balena.io>

### CODE REVIEW

**Comment by lurch** [comment]
> While balenaOS version is a nice addition this does raise the issue of keeping this up to date

I think @imrehg has a page somewhere where all the current balenaOS versions are displayed dynamically? (probably via some JS on the page talking to the balena API or something?)

**Comment by lurch** [comment]
The 1.x version in the
```
Generic | generic | amd64 | 1.2.0
```
entry looks kinda odd? :confused:

**Comment by lurch** [comment]
Should the "Discontinued" header be "Unsupported" instead? :man_shrugging:
Should this also tie in with the "community supported" devices that I saw discussed in some other PR a while ago?

**Comment by garethtdavies** [comment]
> > While balenaOS version is a nice addition this does raise the issue of keeping this up to date
> 
> I think @imrehg has a page somewhere where all the current balenaOS versions are displayed dynamically? (probably via some JS on the page talking to the balena API or something?)

Yes, it is here https://balena-io-playground.github.io/balenaosversions/.

The issue here isn't so much in getting the data (we already have it via the API which this is using) it's more how and when to refresh it. This is a solvable problem, just not one included in this PR 😄

**Comment by garethtdavies** [comment]
> The 1.x version in the
> 
> ```
> Generic | generic | amd64 | 1.2.0
> ```
> 
> entry looks kinda odd? 😕

Yes, I left it in as this is being returned via the API but we can filter this out.

File: tools/device-listing.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Get all devices from device endpoint
```
[the remark this answers is not in the document]
Comment (xginn8) [reply tools/device-listing.py]: What is this script used for?

[the remark this answers is not in the document]
Comment (garethtdavies) [reply tools/device-listing.py]: I just included the link to a gist as a reference, it wasn't supposed to have been included, that was my mistake as lots of files in this PR...

[the remark this answers is not in the document]
Comment (garethtdavies) [reply tools/device-listing.py]: Ah, that wasn't even the latest version just some early testing that accidentally got staged. Removed that file, thanks for the catch.

[the remark this answers is not in the document]
Comment (xginn8) [reply tools/device-listing.py]: No worries, thanks for checking!

**Comment by garethtdavies** [comment]
Related there is a PR here: https://github.com/balena-io/resin-site/pull/944 to update https://www.balena.io/os/docs/supported-boards/. As has been noted previously there are inconsistencies in our arch naming/output.

**Comment by lurch** [comment]
> The issue here isn't so much in getting the data (we already have it via the API which this is using) it's more how and when to refresh it.

I think what I was meaning, was if @imrehg 's page can dynamically get the balenaOS versions from the API, why couldn't / shouldn't the docs page do the same? Or would querying the balena API _every_ time that particular docs-page got loaded, cause too much load-increase on the API server(s) ? :man_shrugging:

**Comment by garethtdavies** [comment]
> I think what I was meaning, was if @imrehg 's page can dynamically get the balenaOS versions from the API, why couldn't / shouldn't the docs page do the same? Or would querying the balena API _every_ time that particular docs-page got loaded, cause too much load-increase on the API server(s) ? 🤷‍♂

Yes, I also made note of how https://www.balena.io/cloud/ uses the same endpoint so that is probably the way to go.

For now, I've stripped this PR to the bare minimum so we can get the updated device listings out there, which would only require an update on a new device listing.

**Comment by garethtdavies** [comment]
I've updated this to pull dynamically from the device endpoint.  I've staged it here https://www.balena-staging.io/docs/reference/hardware/devices/. 

Front-end dev isn't my specialty 😕 so I'd appreciate a review on this file in particular: https://github.com/balena-io/docs/blob/d0f57e404c3f8fa3f05ad758b58579c4e0df6f2e/templates/_devices.html cc @sradevski

**Comment by lurch** [comment]
Will the "Discontinued" section be re-appearing?

**Comment by lurch** [comment]
Tangentially related to this, there's also the download links on https://www.balena.io/os/
(I know this is probably the wrong place to be reporting this, but I dunno what the right place is...)

@garethtdavies By looking at the download URLs, it seems the "Raspberry Pi 3 B/B+" image on that page is `raspberrypi3` and the "Raspberry Pi 4" image is `raspberrypi4-64` - I wonder if that page needs to clarify that the Pi3 image is 32-bit but the Pi4 image is 64-bit?

**Comment by garethtdavies** [comment]
> @garethtdavies By looking at the download URLs, it seems the "Raspberry Pi 3 B/B+" image on that page is `raspberrypi3` and the "Raspberry Pi 4" image is `raspberrypi4-64` - I wonder if that page needs to clarify that the Pi3 image is 32-bit but the Pi4 image is 64-bit?

Thanks, I'll open up a new issue to track this in the morning as I am not sure how this list is generated.

**Pull request merged**